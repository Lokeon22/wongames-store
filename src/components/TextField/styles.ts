import styled, { css, DefaultTheme } from "styled-components"

import { TextFieldProps } from "."

type WrapperProps = Pick<TextFieldProps, "disabled" | "error">

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input} {
      cursor: not-allowed;
      color: ${theme.colors.gray};
    }
  `,

  error: (theme: DefaultTheme) => css`
    span,
    ${Label} {
      color: ${theme.colors.red};
    }

    ${InputWrapper} {
      border-color: ${theme.colors.red};

      svg {
        color: ${theme.colors.red};
      }
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, error }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
    ${!!error && wrapperModifiers.error(theme)}
  `}
`

export const InputWrapper = styled.div<Pick<TextFieldProps, "iconPosition">>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    position: relative;

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    > svg {
      position: relative;
      display: flex;
      order: ${iconPosition === "right" ? 1 : 0};
      right: ${iconPosition === "right" ? "-0.6rem" : "0.6rem"};
      width: 2.2rem;
      color: ${theme.colors.gray};
    }
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    width: 100%;
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmal} 0;
    background: transparent;
    border: 0;
    outline: none;

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;
    }
  `}
`

export const Label = styled.label<Pick<TextFieldProps, "labelColors">>`
  ${({ theme, labelColors }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors[labelColors!]};
  `}
`
