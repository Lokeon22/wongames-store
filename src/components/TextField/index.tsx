import { InputHTMLAttributes, useState } from "react"

import * as S from "./styles"

export type TextFieldProps = {
  onInputChange?: (value: string) => void
  label?: string
  labelFor?: string
  labelColors?: "white" | "black"
  initialValue?: string
  icon?: React.ReactNode
  iconPosition?: "left" | "right"
  disabled?: boolean
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  labelFor,
  labelColors = "white",
  initialValue = "",
  onInputChange,
  icon,
  iconPosition = "left",
  disabled = false,
  error,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  function onChange(e: React.ChangeEvent<HTMLInputElement>) {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInputChange && onInputChange(newValue)
  }

  return (
    <S.Wrapper disabled={disabled} error={error}>
      {label && (
        <S.Label labelColors={labelColors} htmlFor={labelFor}>
          {label}
        </S.Label>
      )}
      <S.InputWrapper iconPosition={iconPosition}>
        {!!icon && icon}
        <S.Input
          type="text"
          value={value}
          onChange={onChange}
          disabled={disabled}
          {...props}
        />
      </S.InputWrapper>
      {!!error && <span>{error}</span>}
    </S.Wrapper>
  )
}

export default TextField
