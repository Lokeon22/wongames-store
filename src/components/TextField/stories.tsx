import { StoryFn, Meta } from "@storybook/react"
import { Email as EmailIcon } from "@styled-icons/material-outlined"
import TextField, { TextFieldProps } from "."

export default {
  title: "Form/TextField",
  component: TextField,
  args: {
    label: "Email",
    labelFor: "Email",
    id: "Email",
    icon: <EmailIcon />,
    initialValue: "",
    placeholder: "gabriel@gmail.com"
  },
  argTypes: {
    icon: {
      control: {
        type: ""
      }
    }
  }
} as Meta

export const Default: StoryFn<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

export const WithError: StoryFn<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

WithError.args = {
  error: "Ocorreu um error"
}
