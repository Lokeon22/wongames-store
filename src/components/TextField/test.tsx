import { screen, waitFor } from "@testing-library/react"
import { renderWithTheme } from "../../utils/tests/helpers"

import TextField from "."
import userEvent from "@testing-library/user-event"

describe("<TextField />", () => {
  it("should render the label", () => {
    renderWithTheme(<TextField label="email" labelFor="email" id="email" />)

    expect(screen.getByLabelText("email")).toBeInTheDocument()
  })

  it("should render without label", () => {
    renderWithTheme(<TextField />)

    expect(screen.queryByLabelText("email")).not.toBeInTheDocument()
  })

  it("should render the input", () => {
    renderWithTheme(<TextField label="email" labelFor="email" id="email" />)

    expect(screen.getByLabelText(/email/i)).toHaveAttribute("type", "text")
  })

  it("should render the placeholder", () => {
    renderWithTheme(<TextField placeholder="hello world" />)

    expect(screen.getByPlaceholderText(/hello world/i)).toBeInTheDocument()
  })

  it("should render the white label by default", () => {
    renderWithTheme(<TextField label="email" labelFor="email" id="email" />)

    expect(screen.getByText(/email/i)).toHaveStyle({
      color: "#FAFAFA"
    })
  })

  it("should render the black label when passed", () => {
    renderWithTheme(
      <TextField
        label="email"
        labelFor="email"
        id="email"
        labelColors="black"
      />
    )

    expect(screen.getByText(/email/i)).toHaveStyle({
      color: "#030517"
    })
  })

  it("should dispatch onChange when input change", async () => {
    const onChange = jest.fn()

    renderWithTheme(
      <TextField
        id="email"
        label="email"
        labelFor="email"
        placeholder="textfield"
        onChange={onChange}
      />
    )

    const element = screen.getByText("email")
    const text = "gabriel@gmail.com"

    userEvent.type(element, text)
    await waitFor(() => {
      expect(onChange).toHaveBeenCalledTimes(text.length)
    })
  })
})
