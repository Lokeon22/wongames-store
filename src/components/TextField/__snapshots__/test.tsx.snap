// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should return error on input 1`] = `
.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 span,
.c0 .c1 {
  color: #FF6347;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c3 svg {
  color: #FF6347;
}

.c4 {
  display: flex;
  align-items: center;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  position: relative;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c4 >svg {
  position: relative;
  display: flex;
  order: 0;
  right: 0.6rem;
  width: 2.2rem;
  color: #8F8F8F;
}

.c6 {
  width: 100%;
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.4rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
}

.c2 {
  font-size: 1.4rem;
  color: #FAFAFA;
}

<div
  class="c0"
  error="error"
>
  <label
    class="c1 c2"
    for="email"
    labelcolors="white"
  >
    email
  </label>
  <div
    class="c3 c4"
    iconposition="left"
  >
    <svg
      aria-hidden="true"
      class="c5"
      data-test="icon"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
      />
    </svg>
    <input
      class="c6"
      id="email"
      type="text"
      value=""
    />
  </div>
  <span>
    error
  </span>
</div>
`;
